# 01. Importing Libraries

In [1]:
#Import Libraries (Step #3)

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02. Importing Data

In [2]:
# Create a string path

path = r'C:\Users\rasmu\Documents\CareerFoundry\A4 Instacart Basket Analysis\02 Data'

In [3]:
# Import Customer data set (Step #3)

df_cust = pd.read_csv(os.path.join(path, 'Orig Data', 'customers.csv'), index_col = False)

# 03. Data Wrangling

In [4]:
df_cust

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [5]:
df_cust.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [6]:
df_cust.shape

(206209, 10)

In [7]:
# Step 4: Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop 
#         columns that don’t add anything to your analysis.

# Checking data types of the columns

df_cust.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [8]:
# Changing the datatype of user_id to a string

df_cust['user_id'] = df_cust['user_id'].astype('str')

In [9]:
# Rechecking the data types

df_cust.dtypes

user_id         object
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [10]:
# Changing column names

df_cust.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [11]:
df_cust.rename(columns = {'n_dependants' : 'Dependants'}, inplace = True)

In [12]:
df_cust.rename(columns = {'fam_status' : 'Marital_status'}, inplace = True)

In [13]:
df_cust.rename(columns = {'STATE' : 'State'}, inplace = True)

In [14]:
# Checking column name changes

df_cust.columns

Index(['user_id', 'First Name', 'Last Name', 'Gender', 'State', 'Age',
       'date_joined', 'Dependants', 'Marital_status', 'income'],
      dtype='object')

# 05. Data Quality & Consistency Checks

In [15]:
# Step 5: Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example,
#         check for and address missing values and duplicates, and convert any mixed-type data.

df_cust.describe()

Age     Dependants         income
count  206209.000000  206209.000000  206209.000000
mean       49.501646       1.499823   94632.852548
std        18.480962       1.118433   42473.786988
min        18.000000       0.000000   25903.000000
25%        33.000000       0.000000   59874.000000
50%        49.000000       1.000000   93547.000000
75%        66.000000       3.000000  124244.000000
max        81.000000       3.000000  593901.000000

Noted that customer ages range from 18 to 81, number of dependents range from 0 to 3, and income ranges from 25903 to 593901. 

In [16]:
# Checking for any missing values

df_cust.isnull().sum()

user_id               0
First Name        11259
Last Name             0
Gender                0
State                 0
Age                   0
date_joined           0
Dependants            0
Marital_status        0
income                0
dtype: int64

Since the First name is the values missing, I didn't forsee removing the values, but this now makes me realize that names are probably not necessary.

In [17]:
# Dropping the First & Last Name columns

df_cust.drop(columns = ['First Name', 'Last Name'])

user_id  Gender           State  Age date_joined  Dependants  \
0        26711  Female        Missouri   48    1/1/2017           3   
1        33890  Female      New Mexico   36    1/1/2017           0   
2        65803    Male           Idaho   35    1/1/2017           2   
3       125935  Female            Iowa   40    1/1/2017           0   
4       130797  Female        Maryland   26    1/1/2017           1   
...        ...     ...             ...  ...         ...         ...   
206204  168073  Female  North Carolina   44    4/1/2020           1   
206205   49635    Male          Hawaii   62    4/1/2020           3   
206206  135902  Female        Missouri   66    4/1/2020           2   
206207   81095  Female      California   27    4/1/2020           1   
206208   80148  Female        New York   55    4/1/2020           1   

       Marital_status  income  
0             married  165665  
1              single   59285  
2             married   99568  
3              single   42049  
4             married   40374  
...               ...     ...  
206204        married  148828  
206205        married  168639  
206206        married   53374  
206207        married   99799  
206208        married   57095  

[206209 rows x 8 columns]

In [18]:
# Rewriting the df_cust with the dropped columns

df_cust = df_cust.drop(columns = ['First Name', 'Last Name'])

In [19]:
# Rechecking for any missing values

df_cust.isnull().sum()

user_id           0
Gender            0
State             0
Age               0
date_joined       0
Dependants        0
Marital_status    0
income            0
dtype: int64

In [20]:
# Checking for Duplicates

df_dups = df_cust[df_cust.duplicated()]

In [21]:
df_dups

Empty DataFrame
Columns: [user_id, Gender, State, Age, date_joined, Dependants, Marital_status, income]
Index: []

This tells me there's no duplicates to be found in the customer data.

In [22]:
# Checking for mixed-type columns

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

No mixed-type columns to report.

# 06. Combing Data

In [23]:
# Importing Dataframe

ords_prods = pd.read_pickle(os.path.join(path, 'Prepd Data', 'ords_prods_merge.pkl'))

In [24]:
# Step 6: Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the 
#         same data type!

# Checking dimensions of dataframe

ords_prods.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True       196.0                1.0   
1                     NaN         True     14084.0                2.0   
2                     NaN         True     12427.0                3.0   
3                     NaN         True     26088.0                4.0   
4                     NaN         True     26405.0                5.0   

   reordered  ... _merge     busiest_day    busiest_days  \
0        0.0  ...   both  Regularly busy  Regularly busy   
1        0.0  ...   both  Regularly busy  Regularly busy   
2        0.0  ...   both  Regularly busy  Regularly busy   
3        0.0  ...   both  Regularly busy  Regularly busy   
4        0.0  ...   both  Regularly busy  Regularly busy   

   busiest_period_of_day max_order      loyalty_flag avg_price spending_flag  \
0         Average Orders        11  Regular customer  6.367797   Low spender   
1         Average Orders        11  Regular customer  6.367797   Low spender   
2         Average Orders        11  Regular customer  6.367797   Low spender   
3         Average Orders        11  Regular customer  6.367797   Low spender   
4         Average Orders        11  Regular customer  6.367797   Low spender   

   order_frequency    frequency_flag  
0             20.0  Regular customer  
1             20.0  Regular customer  
2             20.0  Regular customer  
3             20.0  Regular customer  
4             20.0  Regular customer  

[5 rows x 24 columns]

In [25]:
# Dropping the current _merge column from ords_prods so I can make a new one after the merge with df_cust

ords_prods = ords_prods.drop('_merge', 1)

In [26]:
# Changing the datatype of user_id to a string to match the other dataframe

ords_prods['user_id'] = ords_prods['user_id'].astype('str')

In [27]:
#Checking the drop and change were successful and the data types of the columns

ords_prods.dtypes

order_id                    int64
user_id                    object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
first_order                  bool
product_id                float64
add_to_cart_order         float64
reordered                 float64
product_name               object
aisle_id                  float64
department_id             float64
prices                    float64
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
avg_price                 float64
spending_flag              object
order_frequency           float64
frequency_flag             object
dtype: object

Attempted to merge df_cust with ords_prods on the 'user_id' key...
'df_opc = ords_prods.merge(df_cust, on = 'user_id', how = 'outer', indicator = True)'
...but I received a memory error - 
MemoryError: Unable to allocate 249. MiB for an array with shape (1, 32641268) and data type object

In [28]:
# After making a number of attempts to change the data types of ords_prods variables (turning things into strings, dropping a 
# column), I'm attempting to change the data types of either the int64, float64, or both.

# Finding the max of variables to determine if _32 or _16 datatypes will work better.

ords_prods.max()

order_id                           3421083
user_id                              99999
order_number                           100
orders_day_of_week                       6
order_hour_of_day                       23
days_since_prior_order                30.0
first_order                           True
product_id                         49688.0
add_to_cart_order                    145.0
reordered                              1.0
aisle_id                             134.0
department_id                         21.0
prices                             99999.0
busiest_day                 Regularly busy
busiest_days                  Slowest days
busiest_period_of_day          Most Orders
max_order                              100
loyalty_flag              Regular customer
avg_price                        25005.425
spending_flag                  Low spender
order_frequency                       30.0
frequency_flag            Regular customer
dtype: object

In [29]:
# Changing the datatypes to float16, int16, and category (where applicable)

ords_prods['order_number'] = ords_prods['order_number'].astype('int16')

In [30]:
ords_prods['orders_day_of_week'] = ords_prods['orders_day_of_week'].astype('int16')

In [31]:
ords_prods['days_since_prior_order'] = ords_prods['days_since_prior_order'].astype('float16')

In [32]:
ords_prods['product_id'] = ords_prods['product_id'].astype('float32')

In [33]:
ords_prods['add_to_cart_order'] = ords_prods['add_to_cart_order'].astype('float16')

In [34]:
ords_prods['reordered'] = ords_prods['reordered'].astype('float16')

In [35]:
ords_prods['aisle_id'] = ords_prods['aisle_id'].astype('float16')

In [36]:
ords_prods['department_id'] = ords_prods['department_id'].astype('float16')

In [37]:
ords_prods['product_name'] = ords_prods['product_name'].astype('category')

In [38]:
ords_prods['prices'] = ords_prods['prices'].astype('float32')

In [39]:
ords_prods['avg_price'] = ords_prods['avg_price'].astype('float16')

In [40]:
ords_prods['busiest_day'] = ords_prods['busiest_day'].astype('category')

In [41]:
ords_prods['busiest_days'] = ords_prods['busiest_days'].astype('category')

In [42]:
ords_prods['busiest_period_of_day'] = ords_prods['busiest_period_of_day'].astype('category')

In [43]:
ords_prods['order_hour_of_day'] = ords_prods['order_hour_of_day'].astype('int16')

In [44]:
ords_prods['max_order'] = ords_prods['max_order'].astype('int16')

In [45]:
ords_prods['loyalty_flag'] = ords_prods['loyalty_flag'].astype('category')

In [46]:
ords_prods['spending_flag'] = ords_prods['spending_flag'].astype('category')

In [47]:
ords_prods['order_frequency'] = ords_prods['order_frequency'].astype('float16')

In [48]:
ords_prods['frequency_flag'] = ords_prods['frequency_flag'].astype('category')

In [49]:
#Checking the datatype changes were successful

ords_prods.dtypes

order_id                     int64
user_id                     object
order_number                 int16
orders_day_of_week           int16
order_hour_of_day            int16
days_since_prior_order     float16
first_order                   bool
product_id                 float32
add_to_cart_order          float16
reordered                  float16
product_name              category
aisle_id                   float16
department_id              float16
prices                     float32
busiest_day               category
busiest_days              category
busiest_period_of_day     category
max_order                    int16
loyalty_flag              category
avg_price                  float16
spending_flag             category
order_frequency            float16
frequency_flag            category
dtype: object

After new attempt to merge, getting closer on memory. Now it reads - MemoryError: Unable to allocate 62.3 MiB for an array with shape (1, 32641268) and data type int16

In [50]:
# Attempting to merge the two dataframes

df_opc = ords_prods.merge(df_cust, on = 'user_id', how = 'outer', indicator = True)

The merge finally worked. I tried so many edits and didn't think to use category. 

# 07. Exporting Pickle

In [51]:
# Step 8- Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

df_opc.to_pickle(os.path.join(path,'Prepd Data', 'opc.pkl'))